components/brave_savings/browser/perf_predictor_web_contents_observer.cc

```diff
@@ -0,0 +1,160 @@
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#endif
+#include "components/prefs/pref_service.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_user_data.h"
+#include "content/public/common/resource_type.h"
+#include "brave/common/pref_names.h"
+#include "brave/vendor/brave_performance_predictor/src/third_party_extractor.h"
+#include "brave/vendor/brave_performance_predictor/src/third_parties.h"
+
+using content::ResourceType;
+using base::Singleton;
+using content::WebContents;
+
+namespace brave_perf_predictor {
+
+class ThirdPartyExtractor;
+
+PerfPredictorWebContentsObserver::PerfPredictorWebContentsObserver(
+    content::WebContents* web_contents)
+    : WebContentsObserver(web_contents),
+      tab_id_(SessionTabHelper::IdForTab(web_contents)) {
+  if (!tab_id_.is_valid()) {
+    return;
+  }
+
+  Profile* profile = Profile::FromBrowserContext(
+    web_contents->GetBrowserContext());
+  if (!profile || profile->IsOffTheRecord()) {
+    return;
+  }
+
+  ThirdPartyExtractor* extractor = ThirdPartyExtractor::GetInstance();
+  if (!extractor->is_initialized()) {
+    extractor->load_entities(static_third_party_config);
+  }
+  bandwidth_predictor_ = new BandwidthSavingsPredictor(extractor);
+}
+
+PerfPredictorWebContentsObserver::~PerfPredictorWebContentsObserver() {
+  if (bandwidth_predictor_) {
+    delete bandwidth_predictor_;
+  }
+}
+
+void PerfPredictorWebContentsObserver::DidStartNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle->IsInMainFrame() || handle->IsDownload()) {
+    return;
+  }
+  // Gather prediction of the _previous_ navigation
+  if (navigation_id_ != 0) {
+    RecordSaving();
+  }
+}
+
+void PerfPredictorWebContentsObserver::ReadyToCommitNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle->IsInMainFrame() || handle->IsDownload()) {
+    return;
+  }
+  // Reset predictor state when we're committed to this navigation
+  bandwidth_predictor_->Reset();
+  navigation_id_ = handle->GetNavigationId();
+  VLOG(2) << tab_id_
+    << " committed navigation ID " << navigation_id_
+    << " to " << handle->GetURL().GetContent();
+}
+
+void PerfPredictorWebContentsObserver::DidFinishNavigation(
+    content::NavigationHandle* handle) {
+  if (!handle->IsInMainFrame() ||
+      !handle->HasCommitted() ||
+      handle->IsDownload()) {
+    return;
+  }
+
+  main_frame_url_ = handle->GetURL();
+}
+
+void PerfPredictorWebContentsObserver::RecordSaving() {
+  if (bandwidth_predictor_ && web_contents()) {
+    uint64_t saving = (uint64_t)bandwidth_predictor_->predict();
+    if (saving > 0) {
+      PrefService* prefs = Profile::FromBrowserContext(
```

Comment:
you can use `user_prefs::UserPrefs::Get(browser_context)` to get the pref service